README.rst
comodo_api provides a wrapper around the Comodo SOAP API to allow for simple(r) interaction
with their API.

All responses are in jsend format for simplicity: https://labs.omniti.com/labs/jsend





HISTORY.rst
.. :changelog:

History
-------
0.3.0 (2017-12-21)
++++++++++++++++++
* Move to jsend for return format
* Test failures as well as success

0.2.2 (2017-12-06)
++++++++++++++++++
* Expand tests
* Make returns even more consistent

0.2.1 (2017-12-04)
++++++++++++++++++

* Fix submit to work with new layout

0.2.0 (2017-12-03)
++++++++++++++++++

* Rework code to be more jsony
* Expand documentation and simplify

0.1.0 (2017-11-30)
++++++++++++++++++

* Initial release


tests/conftest.py
import betamax
import os
from betamax_serializers import pretty_json

betamax.Betamax.register_serializer(pretty_json.PrettyJSONSerializer)

customer_login_uri = os.environ.get('CUSTOMER_LOGIN_URI', 'comodo')
login = os.environ.get('LOGIN', 'comodo')
org_id = os.environ.get('ORG_ID', '1234')
password = os.environ.get('PASSWORD', 'fubar')
secret_key = os.environ.get('SECRET_KEY', 'fubar')

with betamax.Betamax.configure() as config:
    config.cassette_library_dir = 'tests/integration/cassettes'
    config.default_cassette_options['serialize_with'] = 'prettyjson'
    config.define_cassette_placeholder('<CUSTOMER_LOGIN_URI>', customer_login_uri)
    config.define_cassette_placeholder('<LOGIN>', login)
    config.define_cassette_placeholder('<ORG_ID>', org_id)
    config.define_cassette_placeholder('<PASSWORD>', password)
    config.define_cassette_placeholder('<SECRET_KEY>', secret_key)


tests/integration/test_session.py
import betamax
import comodo_api
import os
import pytest

customer_login_uri = os.environ.get('CUSTOMER_LOGIN_URI', 'comodo')
login = os.environ.get('LOGIN', 'comodo')
org_id = os.environ.get('ORG_ID', '1234')
password = os.environ.get('PASSWORD', 'fubar')
secret_key = os.environ.get('SECRET_KEY', 'fubar')

# A 'good' working client
@pytest.fixture(scope='class')
def api_client():
    yield comodo_api.ComodoTLSService(api_url='https://hard.cert-manager.com/private/ws/EPKIManagerSSL?wsdl',
                                      customer_login_uri=customer_login_uri, login=login, org_id=org_id,
                                      client_cert_auth=True,
                                      client_public_certificate='/etc/pki/tls/certs/comodo_client.crt',
                                      client_private_key='/etc/pki/tls/private/comodo_client.key',
                                      secret_key=secret_key, password=password)

# A 'bad' or bogus client to generate errors
@pytest.fixture(scope='class')
def bad_api_client():
    yield comodo_api.ComodoTLSService(api_url='https://hard.cert-manager.com/private/ws/EPKIManagerSSL?wsdl',
                                      customer_login_uri='BadURI', login='BadLogin', org_id=1234,
                                      client_cert_auth=False,
                                      secret_key='BadKey', password='BadPassword')

class TestComodoAPI(object):
    def test_get_cert_types(self, api_client):
        recorder = betamax.Betamax(api_client.session)
        with recorder.use_cassette('ComodoAPI_get_cert_types'):
            result = api_client.get_cert_types()

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'success' in result['status']
        assert 'data' in result
        assert 'cert_types' in result['data']
        assert isinstance(result['data']['cert_types'], list)

    def test_get_cert_types_failure(self, bad_api_client):
        recorder = betamax.Betamax(bad_api_client.session)
        with recorder.use_cassette('ComodoAPI_get_cert_types_failure'):
            result = bad_api_client.get_cert_types()

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'error' in result['status']

    def test_revoke(self, api_client):
        recorder = betamax.Betamax(api_client.session)
        with recorder.use_cassette('ComodoAPI_revoke'):
            result = api_client.revoke(cert_id=123456, reason='Revoked for testing')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'success' in result['status']

    def test_revoke_failure(self, bad_api_client):
        recorder = betamax.Betamax(bad_api_client.session)
        with recorder.use_cassette('ComodoAPI_revoke_failure'):
            result = bad_api_client.revoke(cert_id=123456, reason='Revoked for testing')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'error' in result['status']

    def test_collect(self, api_client):
        recorder = betamax.Betamax(api_client.session)
        with recorder.use_cassette('ComodoAPI_collect'):
            result = api_client.collect(cert_id=123456, format_type='X509 PEM Bundle')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'success' in result['status']
        assert 'data' in result
        assert 'certificate' in result['data']
        assert 'certificate_id' in result['data']
        assert 'certificate_status' in result['data']

    def test_collect_failure(self, bad_api_client):
        recorder = betamax.Betamax(bad_api_client.session)
        with recorder.use_cassette('ComodoAPI_collect_failure'):
            result = bad_api_client.collect(cert_id=123456, format_type='X509 PEM Bundle')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'error' in result['status']

    def test_submit(self, api_client):
        csr='''-----BEGIN CERTIFICATE REQUEST-----
        <CSR_DATA>
        -----END CERTIFICATE REQUEST-----
        '''
        recorder = betamax.Betamax(api_client.session)
        with recorder.use_cassette('ComodoAPI_submit'):
            result = api_client.submit(cert_type_name='PlatinumSSL Certificate',
                                       csr=csr, term=3, revoke_password='foo,bar',
                                       server_type='Apache/ModSSL', subject_alt_names='test2.colorado.edu')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'success' in result['status']
        assert 'data' in result
        assert 'certificate_id' in result['data']

    def test_submit_failure(self, bad_api_client):
        csr = '''-----BEGIN CERTIFICATE REQUEST-----
         <CSR_DATA>
         -----END CERTIFICATE REQUEST-----
         '''
        recorder = betamax.Betamax(bad_api_client.session)
        with recorder.use_cassette('ComodoAPI_submit_failure'):
            result = bad_api_client.submit(cert_type_name='PlatinumSSL Certificate',
                                           csr=csr, term=3, revoke_password='foo,bar',
                                           server_type='Apache/ModSSL', subject_alt_names='test2.colorado.edu')

        assert isinstance(result, dict)
        assert 'status' in result
        assert 'error' in result['status']


comodo_api/comodo_api.py
import jsend
from requests import Session
from zeep import Client
from zeep.transports import Transport


class ComodoCA(object):
    """
    Top level class for the Comodo CA. Only very generic 'things' go here.
    """

    format_type = {'X509 PEM Bundle': 0,
                   'X509 PEM Certificate only': 1,
                   'X509 PEM Intermediate certificate only': 2,
                   'PKCS#7 PEM Bundle': 3,
                   }

    formats = {'AOL': 1,
               'Apache/ModSSL': 2,
               'Apache-SSL': 3,
               'C2Net Stronghold': 4,
               'Cisco 3000 Series VPN Concentrator': 33,
               'Citrix': 34,
               'Cobalt Raq': 5,
               'Covalent Server Software': 6,
               'IBM HTTP Server': 7,
               'IBM Internet Connection Server': 8,
               'iPlanet': 9,
               'Java Web Server (Javasoft / Sun)': 10,
               'Lotus Domino': 11,
               'Lotus Domino Go!': 12,
               'Microsoft IIS 1.x to 4.x': 13,
               'Microsoft IIS 5.x and later': 14,
               'Netscape Enterprise Server': 15,
               'Netscape FastTrac': 16,
               'Novell Web Server': 17,
               'Oracle': 18,
               'Quid Pro Quo': 19,
               'R3 SSL Server': 20,
               'Raven SSL': 21,
               'RedHat Linux': 22,
               'SAP Web Application Server': 23,
               'Tomcat': 24,
               'Website Professional': 25,
               'WebStar 4.x and later': 26,
               'WebTen (from Tenon)': 27,
               'Zeus Web Server': 28,
               'Ensim': 29,
               'Plesk': 30,
               'WHM/cPanel': 31,
               'H-Sphere': 32,
               'OTHER': -1,
               }

    status_code = {1: 'Certificate available',
                   2: 'Certificates Attached',
                   0: 'Successful',
                   -10: 'The CSR cannot be decoded!',
                   -11: 'The CSR uses an unsupported algorithm!',
                   -12: 'The CSR has an invalid signature!',
                   -13: 'The CSR uses an unsupported key size!',
                   -14: 'An unknown error occurred!',
                   -16: 'Permission denied!',
                   -20: 'The certificate request has been rejected!',
                   -21: 'The certificate has been revoked!',
                   -22: 'Still awaiting payment!',
                   -31: 'The email is not a valid email.',
                   -32: 'The two phrase should be the same!',
                   -33: 'The Comodo certificate type is invalid!',
                   -34: 'The secret key is invalid!',
                   -35: 'The server type is invalid!',
                   -36: 'The term is invalid for customer type!',
                   -40: 'Invalid ID',
                   -100: 'Invalid authentication data for customer',
                   -101: 'Invalid authentication data for customer Organization',
                   -110: 'Domain is not allowed for customer',
                   -111: 'Domain is not allowed for customer Organization',
                   -120: 'Customer configuration is not allowed the requested action',
                   }


class ComodoTLSService(ComodoCA):
    """
    Class that encapsulates methods to use against Comodo SSL/TLS certificates
    """
    def __init__(self, **kwargs):
        """
        :param string api_url: The full URL for the API server
        :param string customer_login_uri: The URI for the customer login (if your login to the Comodo GUI is at
                https://hard.cert-manager.com/customer/foo/, your login URI is 'foo').
        :param string login: The login user
        :param string org_id: The organization ID
        :param string password: The API user's password
        :param string secret_key: The API user's secret key
        :param bool client_cert_auth: Whether to use client certificate authentication
        :param string client_public_certificate: The path to the public key if using client cert auth
        :param string client_private_key: The path to the private key if using client cert auth
        """
        # Using get for consistency and to allow defaults to be easily set
        self.api_url = kwargs.get('api_url')
        self.customer_login_uri = kwargs.get('customer_login_uri')
        self.login = kwargs.get('login')
        self.org_id = kwargs.get('org_id')
        self.password = kwargs.get('password')
        self.secret_key = kwargs.get('secret_key')
        self.client_cert_auth = kwargs.get('client_cert_auth')
        self.session = Session()
        self.transport = Transport(session=self.session)
        self.client = Client(self.api_url, transport=self.transport)
        # Because Comodo is crap at designing APIs (in my opinion) we have to get the wsdl
        # then modify the transport to use client certs after that.
        if self.client_cert_auth:
            self.client_public_certificate = kwargs.get('client_public_certificate')
            self.client_private_key = kwargs.get('client_private_key')
            self.session.cert = (self.client_public_certificate, self.client_private_key)
        self.type_factory = self.client.type_factory('ns0')
        self.auth = self.type_factory.AuthData()
        self.auth.login = self.login
        self.auth.password = self.password
        self.auth.customerLoginUri = self.customer_login_uri

    def _create_error(self, status_code):
        """
        Construct an error message in jsend format.

        :param int status_code: The status code to translate into an error message
        :return: A dictionary in jsend format with the error and the code
        :rtype: dict
        """

        return jsend.error(message=ComodoCA.status_code[status_code], code=status_code)

    def get_cert_types(self):
        """
        Collect the certificate types that are available to the customer.

        :return: A list of dictionaries of certificate types
        :rtype: list
        """
        result = self.client.service.getCustomerCertTypes(authData=self.auth)

        if result.statusCode == 0:
            return jsend.success({'cert_types': result.types})
        else:
            return self._create_error(result.statusCode)

    def collect(self, cert_id, format_type):
        """
        Poll for certificate availability after submission.

        :param int cert_id: The certificate ID
        :param str format_type: The format type to use (example: 'X509 PEM Certificate only')
        :return: The certificate_id or the certificate depending on whether the certificate is ready (check status code)
        :rtype: dict
        """

        result = self.client.service.collect(authData=self.auth, id=cert_id,
                                             formatType=ComodoCA.format_type[format_type])

        # The certificate is ready for collection
        if result.statusCode == 2:
            return jsend.success({'certificate': result.SSL.certificate, 'certificate_status': 'issued',
                                  'certificate_id': cert_id})
        # The certificate is not ready for collection yet
        elif result.statusCode == 0:
            return jsend.fail({'certificate_id': cert_id, 'certificate': '', 'certificate_status': 'pending'})
        # Some error occurred
        else:
            return self._create_error(result.statusCode)

    def revoke(self, cert_id, reason=''):
        """
        Revoke a certificate.

        :param int cert_id: The certificate ID
        :param str reason: Reason for revocation (up to 256 characters), can be blank: ''
        :return: The result of the operation, 'Successful' on success
        :rtype: dict
        """
        result = self.client.service.revoke(authData=self.auth, id=cert_id, reason=reason)

        if result == 0:
            return jsend.success()
        else:
            return self._create_error(result)

    def submit(self, cert_type_name, csr, revoke_password, term, subject_alt_names='',
               server_type='OTHER'):
        """
        Submit a certificate request to Comodo.

        :param string cert_type_name: The full cert type name (Example: 'PlatinumSSL Certificate') the supported
                                      certificate types for your account can be obtained with the
                                      get_cert_types() method.
        :param string csr: The Certificate Signing Request (CSR)
        :param string revoke_password: A password for certificate revocation
        :param int term: The length, in years, for the certificate to be issued
        :param string subject_alt_names: Subject Alternative Names separated by a ",".
        :param string server_type: The type of server for the TLS certificate e.g 'Apache/ModSSL' full list available in
                                   ComodoCA.server_type (Default: OTHER)
        :return: The certificate_id and the normal status messages for errors.
        :rtype: dict
        """
        cert_types = self.get_cert_types()

        # If collection of cert types fails we simply pass the error back.
        if cert_types['status'] == 'error':
            return cert_types

        # We do this because we need to pass the entire cert type definition back to Comodo
        # not just the name.
        for cert_type in cert_types['data']['cert_types']:
            if cert_type.name == cert_type_name:
                cert_type_def = cert_type

        result = self.client.service.enroll(authData=self.auth, orgId=self.org_id, secretKey=self.secret_key,
                                            csr=csr, phrase=revoke_password, subjAltNames=subject_alt_names,
                                            certType=cert_type_def, numberServers=1,
                                            serverType=ComodoCA.formats[server_type], term=term, comments='')

        # Anything greater than 0 is the certificate ID
        if result > 0:
            return jsend.success({'certificate_id': result})
        # Anything else is an error
        else:
            return self._create_error(result)
